utils/tracefs.c

```diff
@@ -1,67 +1,59 @@
 #include <fcntl.h>
+#include <linux/magic.h>
+#include <mntent.h>
 #include <stdio.h>
 #include <string.h>
+#include <sys/vfs.h>
 
 #include "utils/tracefs.h"
 #include "utils/utils.h"
 
-#define PROC_MOUNTINFO "/proc/self/mountinfo"
+#define TRACEFS_DIR_PATH "/sys/kernel/tracing"
 
 static char *TRACING_DIR = NULL;
 
 static bool find_tracing_dir(void)
 {
-	FILE *fp;
-	char *line = NULL, fs_type[NAME_MAX], mount_point[PATH_MAX];
-	static char debugfs_suffix[] = "tracing";
-	bool debugfs_found = false;
-	size_t len;
-
-	if (TRACING_DIR)
-		return false;
-
-	fp = fopen(PROC_MOUNTINFO, "r");
-	if (fp == NULL)
-		return false;
-
-	while (getline(&line, &len, fp) > 0) {
-		/*
-		 * /proc/<pid>/mountinfo format:
-		 * 36 35 98:0 /mnt1 /mnt2 rw,noatime master:1 - ext3 .... ....
-		 * (1)(2)(3)   (4)   (5)      (6)      (7)   (8) (9) (10) (11)
-		 *                mount_point                  fs_type
-		 *
-		 * (9) is the file system type, (5) is the mount point relative
-		 * to self's root directory.
-		 */
-		sscanf(line, "%*i %*i %*u:%*u %*s %s %*s %*s - %s %*s %*s\n", mount_point, fs_type);
-
-		if (!strcmp(fs_type, "tracefs")) {
-			/* discard previously kept debugfs tracing dir */
-			if (TRACING_DIR)
-				free(TRACING_DIR);
-			xasprintf(&TRACING_DIR, "%s", mount_point);
-			pr_dbg2("Found tracefs at %s\n", mount_point);
-			pr_dbg2("Use %s as TRACING_DIR\n", TRACING_DIR);
-			return true;
-		}
-
-		if (!strcmp(fs_type, "debugfs")) {
-			xasprintf(&TRACING_DIR, "%s/%s", mount_point, debugfs_suffix);
-			pr_dbg2("Found debugfs at %s\n", mount_point);
-			pr_dbg2("Keep searching for tracefs...\n");
-			debugfs_found = true;
-		}
+	struct statfs fs;
+	bool found = false;
+
+	if (!statfs(TRACEFS_DIR_PATH, &fs) && fs.f_type == TRACEFS_MAGIC) {
+		xasprintf(&TRACING_DIR, "%s", TRACEFS_DIR_PATH);
+		found = true;
+	}
+	else {
+		struct mntent *ent;
+		FILE *fp;
+
+		do {
+			fp = setmntent("/proc/mounts", "r");
+			if (fp == NULL) {
+				break;
+			}
+
+			while (found != true && NULL != (ent = getmntent(fp))) {
+				if (!strcmp(ent->mnt_fsname, "tracefs")) {
+					xasprintf(&TRACING_DIR, "%s", ent->mnt_dir);
+					found = true;
+				}
+				else if (!strcmp(ent->mnt_fsname, "debugfs")) {
+					xasprintf(&TRACING_DIR, "%s/tracing", ent->mnt_dir);
+					found = true;
+				}
+			}
+
+			endmntent(fp);
+		} while (0);
```

Comment:
First, I wanted to go down without executing the `while() { ... }` and `endmntent()` below when the result of the `setmntent()` function is NULL.

Second, I didn't want to keep the `if ~ else` statements too long. If the `if ~ else` statement is too long or deep it becomes hard to read. so I wanted to use `do ~ while(0)` which is used by the Linux kernel to avoid this.

I thought this was best, but if there is a better alternative, I'm open to it.

Reply:
I think you can split the check for TRACEFS_DIR_PATH into a separate commit and focus on the mntent change here.  Then it can be like this.

```c
	fp = setmntent("/proc/mounts", "r");
	if (fp == NULL)
		return false;

	while ((ent = getmntent(fp)) != NULL) {
		if (strcmp(ent->mnt_fsname, "tracefs"))
			continue;

		xasprintf(&TRACING_DIR, "%s", ent->mnt_dir);
		found = true;
		break;
	}
	endmntent(fp);
	return found;
```